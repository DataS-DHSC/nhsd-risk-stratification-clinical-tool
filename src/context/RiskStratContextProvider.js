/* eslint-disable max-len */
/* eslint-disable react/prop-types */
import React, { useState } from 'react';

const RiskStratContextProvider = ({
  children,
  RiskStratContext,
  initialState,
}) => {
  let initialContextState = null;
  const isRunningClientSide = typeof window !== 'undefined';
  if (isRunningClientSide) {
    // we only want to use the context state if this code is running client side
    // when javascript is disabled, this code will run server side
    // in that case, we want to return null for the bmaData to force pages to use POST prevData state rather than the context state
    initialContextState = {
      ...initialState,
    };
  }

  const [contextData, setContextData] = useState(initialContextState);

  const setField = (field, value) => {
    setContextData({ ...contextData, [field]: value });
  };

  const resetContext = () => {
    setContextData({ ...initialState });
  };

  return (
    <RiskStratContext.Provider value={[contextData, setField, resetContext]}>
      {children}
    </RiskStratContext.Provider>
  );
};

export default RiskStratContextProvider;
