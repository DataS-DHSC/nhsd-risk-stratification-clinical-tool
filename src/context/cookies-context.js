import { createContext, useContext, useState } from 'react';
import PropTypes from 'prop-types';

function parseCookie(str) {
  switch (str) {
    case 'true':
      return true;
    case 'false':
      return false;
    case undefined:
      return undefined;
    default:
      // eslint-disable-next-line no-console
      console.error(`Unable to parse opt-in value ${str}`);
      return undefined;
  }
}

export const CookiesContext = createContext({
  optIn: undefined,
  setOptIn: () => {
    throw Error(
      'setOptIn function not defined, make sure to use context provider'
    );
  },
  hideSuccessBanner: () => {
    throw Error(
      'hideSuccessBanner function not defined, make sure to use context provider'
    );
  },
  showSuccessBanner: () => {
    throw Error(
      'showSuccessBanner function not defined, make sure to use context provider'
    );
  },
  isSuccessBannerVisible: undefined,
  isSelectionBannerVisible: () => {
    throw Error(
      'isSelectionBannerVisible function not defined, make sure to use context provider'
    );
  },
});

export const useCookiesContext = () => useContext(CookiesContext);

export default function CookiesContextProvider({ children, getCookie }) {
  const [optIn, setOptIn] = useState(parseCookie(getCookie('opt-in')));

  const [isSuccessBannerVisible, setIsSuccessBannerVisible] = useState(false);
  const hideSuccessBanner = () => setIsSuccessBannerVisible(false);
  const showSuccessBanner = () => setIsSuccessBannerVisible(true);

  const isSelectionBannerVisible = () => optIn === undefined;

  return (
    <CookiesContext.Provider
      value={{
        optIn,
        setOptIn,
        hideSuccessBanner,
        showSuccessBanner,
        isSuccessBannerVisible,
        isSelectionBannerVisible,
      }}
    >
      {children}
    </CookiesContext.Provider>
  );
}

CookiesContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
  getCookie: PropTypes.func.isRequired,
};
