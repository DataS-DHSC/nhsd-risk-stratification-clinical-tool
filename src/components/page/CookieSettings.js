/* eslint-disable jsx-a11y/label-has-associated-control */
import { addMonths } from 'date-fns';
import { useRouter } from 'next/router';
import React, { useState } from 'react';
import cookie from 'react-cookies';
import { useCookiesContext } from '../../context/cookies-context';
import {
  boolToAnalyticsInputvalue,
  analyticsInputValueToBool,
} from '../../utils/cookieSettingsUtils';
import { ROUTE_COOKIES_CONFIRMATION } from '../../constants/routes';

function CookieSettings() {
  const router = useRouter();

  const { optIn, setOptIn } = useCookiesContext();

  const [analyticsCookiesInput, setAnalyticsCookiesInput] = useState(
    boolToAnalyticsInputvalue(optIn)
  );

  function handleSubmit(evt) {
    evt.preventDefault();
    const defaultCookieSettings = {
      path: '/',
      sameSite: 'Strict',
    };
    cookie.save(
      'opt-in',
      `${analyticsInputValueToBool(analyticsCookiesInput)}`,
      {
        ...defaultCookieSettings,
        expires: addMonths(new Date(), 12),
      }
    );
    setOptIn(analyticsInputValueToBool(analyticsCookiesInput));
    router.push(ROUTE_COOKIES_CONFIRMATION);
  }

  return (
    <body>
      <div className="nhsuk-grid-row">
        <div className="nhsuk-grid-column-two-thirds">
          <h1 className="nhsuk-heading-l">Change your cookie settings</h1>
          <noscript>
            <p className="nhsuk-body">
              We cannot change your cookie settings at the moment because
              JavaScript is not running in your browser. To fix this, try:
            </p>
            <ul className="nhsuk-list nhsuk-list--bullet">
              <li>turning on JavaScript in your browser settings</li>
              <li>reloading this page</li>
            </ul>
          </noscript>

          <form onSubmit={handleSubmit} method="post" noValidate>
            <div className="nhsuk-form-group">
              <p className="nhsuk-body">
                Let us know which cookies we can use. We&apos;ll use a cookie to
                save your settings.
              </p>
              <fieldset className="nhsuk-fieldset">
                <legend className="nhsuk-fieldset__legend nhsuk-fieldset__legend--s">
                  Cookies that measure website use (analytics cookies)
                </legend>
                <p>
                  These cookies store information about how you use our website,
                  such as the links you click on. We use them to help make our
                  website better.
                </p>
                <div className="nhsuk-radios">
                  <div className="nhsuk-radios__item">
                    <input
                      className="nhsuk-radios__input"
                      id="analytics-cookies-yes"
                      name="analytics-cookies"
                      type="radio"
                      value="yes"
                      onChange={(evt) => {
                        setAnalyticsCookiesInput(evt.target.value);
                      }}
                      checked={analyticsCookiesInput === 'yes'}
                    />
                    <label
                      className="nhsuk-label nhsuk-radios__label"
                      htmlFor="analytics-cookies-yes"
                    >
                      Yes
                    </label>
                  </div>
                  <div className="nhsuk-radios__item">
                    <input
                      className="nhsuk-radios__input"
                      id="analytics-cookies-no"
                      name="analytics-cookies"
                      type="radio"
                      value="no"
                      onChange={(evt) => {
                        setAnalyticsCookiesInput(evt.target.value);
                      }}
                      checked={analyticsCookiesInput === 'no'}
                    />
                    <label
                      htmlFor="analytics-cookies-no"
                      className="nhsuk-label nhsuk-radios__label"
                    >
                      No
                    </label>
                  </div>
                </div>
              </fieldset>
            </div>

            <button
              type="submit"
              className="nhsuk-button"
              data-module="nhsuk-button"
              id="save-cookie-settings"
            >
              Save cookie settings
            </button>
          </form>
        </div>
      </div>
    </body>
  );
}

export default CookieSettings;
