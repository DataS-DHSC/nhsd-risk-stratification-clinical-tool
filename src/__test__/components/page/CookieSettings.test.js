import React from 'react';
import { render } from '@testing-library/react';
import CookieSettings from '../../../components/page/CookieSettings';
import { CookiesContext } from '../../../context/cookies-context';

const customRender = (ui, { providerProps, ...renderOptions }) =>
  render(
    <CookiesContext.Provider {...providerProps}>{ui}</CookiesContext.Provider>,
    renderOptions
  );

describe('CookieSettings', () => {
  it('should match snapshot', () => {
    const providerProps = {
      value: { optIn: true, setOptIn: () => {} },
    };
    const screen = customRender(<CookieSettings />, { providerProps });
    expect(screen.asFragment()).toMatchSnapshot('CookiesAcceptedOnLoad');
  });

  it('should match snapshot', () => {
    const providerProps = {
      value: { optIn: false, setOptIn: () => {} },
    };
    const screen = customRender(<CookieSettings />, { providerProps });
    expect(screen.asFragment()).toMatchSnapshot('CookiesRejectedOnLoad');
  });
});
