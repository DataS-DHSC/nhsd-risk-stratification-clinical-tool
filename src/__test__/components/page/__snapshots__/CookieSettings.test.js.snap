// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieSettings should match snapshot: CookiesAcceptedOnLoad 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1
        class="nhsuk-heading-l"
      >
        Change your cookie settings
      </h1>
      <noscript />
      <form
        method="post"
        novalidate=""
      >
        <div
          class="nhsuk-form-group"
        >
          <p
            class="nhsuk-body"
          >
            Let us know which cookies we can use. We'll use a cookie to save your settings.
          </p>
          <fieldset
            class="nhsuk-fieldset"
          >
            <legend
              class="nhsuk-fieldset__legend nhsuk-fieldset__legend--s"
            >
              Cookies that measure website use (analytics cookies)
            </legend>
            <p>
              These cookies store information about how you use our website, such as the links you click on. We use them to help make our website better.
            </p>
            <div
              class="nhsuk-radios"
            >
              <div
                class="nhsuk-radios__item"
              >
                <input
                  checked=""
                  class="nhsuk-radios__input"
                  id="analytics-cookies-yes"
                  name="analytics-cookies"
                  type="radio"
                  value="yes"
                />
                <label
                  class="nhsuk-label nhsuk-radios__label"
                  for="analytics-cookies-yes"
                >
                  Yes
                </label>
              </div>
              <div
                class="nhsuk-radios__item"
              >
                <input
                  class="nhsuk-radios__input"
                  id="analytics-cookies-no"
                  name="analytics-cookies"
                  type="radio"
                  value="no"
                />
                <label
                  class="nhsuk-label nhsuk-radios__label"
                  for="analytics-cookies-no"
                >
                  No
                </label>
              </div>
            </div>
          </fieldset>
        </div>
        <button
          class="nhsuk-button"
          data-module="nhsuk-button"
          id="save-cookie-settings"
          type="submit"
        >
          Save cookie settings
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CookieSettings should match snapshot: CookiesRejectedOnLoad 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1
        class="nhsuk-heading-l"
      >
        Change your cookie settings
      </h1>
      <noscript />
      <form
        method="post"
        novalidate=""
      >
        <div
          class="nhsuk-form-group"
        >
          <p
            class="nhsuk-body"
          >
            Let us know which cookies we can use. We'll use a cookie to save your settings.
          </p>
          <fieldset
            class="nhsuk-fieldset"
          >
            <legend
              class="nhsuk-fieldset__legend nhsuk-fieldset__legend--s"
            >
              Cookies that measure website use (analytics cookies)
            </legend>
            <p>
              These cookies store information about how you use our website, such as the links you click on. We use them to help make our website better.
            </p>
            <div
              class="nhsuk-radios"
            >
              <div
                class="nhsuk-radios__item"
              >
                <input
                  class="nhsuk-radios__input"
                  id="analytics-cookies-yes"
                  name="analytics-cookies"
                  type="radio"
                  value="yes"
                />
                <label
                  class="nhsuk-label nhsuk-radios__label"
                  for="analytics-cookies-yes"
                >
                  Yes
                </label>
              </div>
              <div
                class="nhsuk-radios__item"
              >
                <input
                  checked=""
                  class="nhsuk-radios__input"
                  id="analytics-cookies-no"
                  name="analytics-cookies"
                  type="radio"
                  value="no"
                />
                <label
                  class="nhsuk-label nhsuk-radios__label"
                  for="analytics-cookies-no"
                >
                  No
                </label>
              </div>
            </div>
          </fieldset>
        </div>
        <button
          class="nhsuk-button"
          data-module="nhsuk-button"
          id="save-cookie-settings"
          type="submit"
        >
          Save cookie settings
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
